### PR TITLE
CORE-1391 UInt truncating cast

### PR DESCRIPTION
**Pull request by erer1243**
\+ changed UIntTy to a real type. Did not yet change stdlib repr of uintty to an enum or something else.

### CODE REVIEW

File: hs/src/Reach/AST/Base.hs [hunk 1]
```diff
@@ -315,11 +315,10 @@ type SLPart = B.ByteString
 render_sp :: SLPart -> Doc
 render_sp = viaShow
 
-type UIntTy = Bool
-uintWord :: UIntTy
-uintWord = False
-uint256 :: UIntTy
-uint256 = True
+data UIntTy = UI_Word | UI_256 deriving (Eq, Generic, NFData, Ord, Show)
```
Comment (jeapostrophe) [inline hs/src/Reach/AST/Base.hs:318]: Gross formatting!

File: hs/src/Reach/Backend/JS.hs [hunk 2]
```diff
@@ -286,7 +286,7 @@ jsPrimApply = \case
   PGE t -> jsApply_ui t "stdlib.ge"
   PGT t -> jsApply_ui t "stdlib.gt"
   SQRT t -> jsApply_ui t "stdlib.sqrt"
-  UCAST x y -> \a -> jsApply "stdlib.cast" $ [ jsUIntTy x, jsUIntTy y ] <> a
+  UCAST dom rng trunc -> \a -> jsApply "stdlib.cast" $ [ jsUIntTy dom, jsUIntTy rng ] <> a <> [ jsBool trunc ]
```
Comment (jeapostrophe) [inline hs/src/Reach/Backend/JS.hs:289]: Make sure you increment the version number every time you expect something different from the stdlib

File: hs/src/Reach/Backend/JS.hs [hunk 3]
```diff
@@ -270,7 +270,7 @@ jsDigest :: AppT [DLArg]
 jsDigest as = jsApply "stdlib.digest" <$> jsContractAndVals as
 
 jsUIntTy :: UIntTy -> Doc
-jsUIntTy = jsBool
+jsUIntTy t = jsBool $ if t == UI_Word then False else True -- TODO change stdlib repr
```
Comment (jeapostrophe) [inline hs/src/Reach/Backend/JS.hs:273]: I think you should do it

File: hs/src/Reach/Eval/Core.hs [hunk 4]
```diff
@@ -2611,6 +2611,12 @@ warnInteractType = \case
 evalPrim :: SLPrimitive -> [SLSVal] -> App SLSVal
 evalPrim p sargs =
   case p of
+    SLPrim_castOrTrunc to -> do
+      let (trunc, sargs') =
+              case sargs of
+                [i, (_, SLV_Bool _ trunc_)] -> (trunc_, [i])
+                _ -> (False, sargs)
```
Comment (jeapostrophe) [inline hs/src/Reach/Eval/Core.hs:2618]: This code can be cleaner. I think it should be like `SLPrim_Token_Burn` and then after do a `fromMaybe False mtrunc`

File: js/stdlib/ts/shared_impl.ts [hunk 5]
```diff
@@ -718,10 +718,12 @@ export const makeArith = (m:BigNumber): Arith => {
   type BNOp2 = 'add'|'sub'|'mod'|'mul'|'div'|'and'|'or'|'xor';
   const doBN2 = (f:BNOp2, a:BigNumber, b:BigNumber) => a[f](b);
   const getCheck = (w:UIntTy) => w ? checkB : checkM;
-  const cast = (from:UIntTy, to:UIntTy, x:num): BigNumber => {
+  const cast = (from:UIntTy, to:UIntTy, x:num, trunc:boolean = false): BigNumber => {
```
Comment (jeapostrophe) [inline js/stdlib/ts/shared_impl.ts:721]: Don't default

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: hs/src/Reach/Backend/JS.hs (hunk 2 above)
Comment (erer1243) [reply hs/src/Reach/Backend/JS.hs:289]: Which version number?

Comment (jeapostrophe) [reply hs/src/Reach/Backend/JS.hs:289]: `reachBackendVersion`

Comment (jeapostrophe) [reply hs/src/Reach/Backend/JS.hs:289]: The JS code will need to be updated to match it

File: examples/overview/build/index.main.mjs [hunk 6]
```diff
@@ -1,8 +1,8 @@
-// Automatically generated with Reach 0.1.10 (f1275a40*)
```
Comment (jeapostrophe) [inline examples/overview/build/index.main.mjs:1]: Use `scripts/restore-build` to not commit this

**Review by jeapostrophe — CHANGES_REQUESTED** [review]
Whenever you change the version, you have to update the changelog to say the compiled objects are different, sorry

File: hs/src/Reach/Backend/JS.hs [hunk 7]
```diff
@@ -270,7 +270,7 @@ jsDigest :: AppT [DLArg]
 jsDigest as = jsApply "stdlib.digest" <$> jsContractAndVals as
 
 jsUIntTy :: UIntTy -> Doc
-jsUIntTy = jsBool
+jsUIntTy t = if t == UI_Word then "stdlib.T_UInt" else "stdlib.T_UInt256"
```
[the remark this answers is not in the document]
Comment (jeapostrophe) [reply hs/src/Reach/Backend/JS.hs:273]: Something went wrong when you were removing the merge conflicts, because you undid this change

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

**Pull request merged**